### DISCUSSION TITLE
Incompatibility between Ember with the native types and `ts-reset`

### DISCUSSION THREAD

**Discussion by boris-petrov** [typed-ember/glint#644] [Q&A]
[Repo](https://github.com/boris-petrov/glint-bug).

Check `tests/types/foo.ts`. It gives a strange error:

```
tests/types/foo.ts:11:7 - error TS2345: Only primitive values and certain DOM objects (see `ContentValue` in `@glint/template`) are usable as top-level template content.
  Argument of type 'UnmetExpectation<"Expected type to be `number`, but got", unknown>' is not assignable to parameter of type 'ContentValue'.

11       {{@expectTypeOf number @to.beNumber}}
         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

If you remove `import '@total-typescript/ts-reset/filter-boolean';` from `types/global.d.ts`, it works fine. It also seems to work fine using the `@types/ember__*` packages instead of the "native" types.

Any ideas why that happens?

**Comment by dfreeman** [comment]
In general I'd be _super_ wary of using libraries like `ts-reset`—though not because the changes it makes aren't nice/reasonable/useful!

The problem is cases exactly like this: monkeypatching the built-in types can have unintended consequences to other libraries. Any time TypeScript itself makes even very small changes to the built-in types, it tends to break many libraries across the ecosystem that were implicitly relying on the prior types. The TS maintainers will often undertake a big PR (or series of PRs) to DefinitelyTyped to account for such changes, but libraries that ship their own types will frequently have to issue bugfix releases because of them.

If you layer on additional tweaks to the standard types from third-party sources like this, it quickly becomes a minefield for maintainers to try and accommodate in projects that are highly sensitive to the details of the type system, like `expect-type` or the Ember types or Glint itself.

**Reply by boris-petrov** [reply]
Thanks for the discussion! I find it strange in this particular case that such a problem occurs. I guess you've seen what this specific "reset" does - it just adds an overload for `Array.filter` and with a very specific argument. Why would that break the world? Also, what do you suggest? Not add that rule? That's what I've done, it's just that `filter(Boolean)` is a nice shortcut to filter-out `null` values... but oh well, if you think there is nothing to be done?
